.github/workflows/latest_deps.yml

```diff
@@ -0,0 +1,160 @@
+# People who are freshly `pip install`ing from the wheels we publish to PyPI will pull
+# in the latest versions of all dependencies. We used to do this in our CI runs, which
+# would detect any incompatibilities due to new package releases. Now that we use the
+# locked poetry environment, we don't have advance notice of these breakages.defaults:
+#
+# To plug that hole, this job is a canary which periodically
+# - checks out develop,
+# - pip installs from source (building a wheel locally using `poetry-core`,
+#   and `pip install`ing from that), and
+# - runs test suites against that checkout.
+#
+# Based on the twisted trunk CI job.
+
+name: Latest dependencies
+
+on:
+  schedule:
+    - cron: 0 7 * * *
+  pull_request:
+    branches: [dmr/poetry-wheel-ci]
+  push:
+    branches: [dmr/poetry-wheel-ci]
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  # TODO(dmr): I'm not sure if we want to run mypy here, nor if mypy failures should
+  # merit an issue.
+  mypy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      # The dev dependencies aren't exposed in the wheel metadata (at least with current
+      # poetry-core versions). So we'll have to be dirty: `poetry install` to get the
+      # dev deps, then upgrade them in situ with pip.
+      - uses: matrix-org/setup-python-poetry@v1
+        with:
+          python-version: "3.x"
+      # Lacking a `pip install --upgrade-literally-everything` option, concoct our own.
+      - run: "poetry run pip freeze | cut -d' ' -f 1 | xargs pip install --upgrade --upgrade-strategy eager"
```

Comment:
I'm unsure if this would happen or not, but... if Synapse depends on Twisted which depends on hyperlink, will this cause the latest version of hyperlink to be installed even if Twisted said < 5.0 or something? (This is a bad example since Twisted only sets a minimum version for hyperlink, but hopefully you understand what I'm getting at.)

Reply:
(Also, I have no other thoughts on how to do this, so maybe this is the best we can do!)

Reply:
> will this cause the latest version of hyperlink to be installed even if Twisted said < 5.0 or something

I don't think this will happen. I experimented with `flake8` as an example, which has an upper bound on `importlib-metadata` under Python 3.7.

```
$ grep importlib /tmp/env/lib/python3.7/site-packages/flake8-4.0.1.dist-info/METADATA 
Requires-Dist: importlib-metadata (<4.3) ; python_version < "3.8"
```

I found that:
- if I installed `importlib-metadata` and `flake8` in the same `pip install` invocation, then pip always selected importlib-metadata 4.2.0. This was the case even if I passed both upgrade flags.
- if I `pip install importlib-metadata` in a clean venv I got importlib-metadata 4.11.3. Installing `flake8` on top of that downgraded importlib-metadata to 4.2.0.
- Installing the other way round and forcing an upgrade of `importlib-metadata` does ignore the upper bound though:
    ```
    $ pip install flake8 --upgrade --upgrade-strategy eager
    ...
    Installing collected packages: importlib-metadata, flake8
    Successfully installed flake8-4.0.1 importlib-metadata-4.2.0
    WARNING: You are using pip version 21.2.3; however, version 22.0.4 is available.
    You should consider upgrading via the '/tmp/env/bin/python3.7 -m pip install --upgrade pip' command.

    $ pip install importlib-metadata --upgrade --upgrade-strategy eager
    Requirement already satisfied: importlib-metadata in ./env/lib/python3.7/site-packages (4.2.0)
    Collecting importlib-metadata
      Using cached importlib_metadata-4.11.3-py3-none-any.whl (18 kB)
    Requirement already satisfied: zipp>=0.5 in ./env/lib/python3.7/site-packages (from importlib-metadata) (3.8.0)
    Requirement already satisfied: typing-extensions>=3.6.4 in ./env/lib/python3.7/site-packages (from importlib-metadata) (4.2.0)
    Installing collected packages: importlib-metadata
      Attempting uninstall: importlib-metadata
        Found existing installation: importlib-metadata 4.2.0
        Uninstalling importlib-metadata-4.2.0:
          Successfully uninstalled importlib-metadata-4.2.0
    ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
    flake8 4.0.1 requires importlib-metadata<4.3; python_version < "3.8", but you have importlib-metadata 4.11.3 which is incompatible.
    Successfully installed importlib-metadata-4.11.3
    WARNING: You are using pip version 21.2.3; however, version 22.0.4 is available.
    You should consider upgrading via the '/tmp/env/bin/python3.7 -m pip install --upgrade pip' command.
    ```
  Pip does at least warn us of the incompatibility.

So as long as the `xargs` invocation only makes one call to `pip install`, I think we're safe.

We could enforce this by passing `-n 1000 -x` to `xargs`, which will make exactly one invocation of `pip install` with up to 1000 package names. (At present we have 126 total dependencies, according to `poetry run pip freeze | wc -l`.) Might be overkill?

Reply:
Having said all that... I think on reflection this would be simpler if I
- used `poetry export --extras all --dev` to get pinned package versions, rather than `poetry install; pip freeze`.